cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ClusterCa.java

```diff
@@ -285,79 +292,74 @@ protected boolean hasCaCertGenerationChanged() {
 
+        // Temp files used when we need to generate new certificates
         File brokerCsrFile = Files.createTempFile("tls", "broker-csr").toFile();
         File brokerKeyFile = Files.createTempFile("tls", "broker-key").toFile();
         File brokerCertFile = Files.createTempFile("tls", "broker-cert").toFile();
         File brokerKeyStoreFile = Files.createTempFile("tls", "broker-p12").toFile();
 
-        int replicasInNewSecret = Math.min(replicasInSecret, replicas);
-        Map<String, CertAndKey> certs = new HashMap<>(replicasInNewSecret);
-        // copying the minimum number of certificates already existing in the secret
-        // scale up -> it will copy all certificates
-        // scale down -> it will copy just the requested number of replicas
-        for (int i = 0; i < replicasInNewSecret; i++) {
-            String podName = podNameFn.apply(i);
-            LOGGER.debugCr(reconciliation, "Certificate for {} already exists", podName);
-            Subject subject = subjectFn.apply(i);
-
-            CertAndKey certAndKey;
-
-            if (isNewVersion(secret, podName)) {
-                certAndKey = asCertAndKey(secret, podName);
-            } else {
-                // coming from an older operator version, the secret exists but without keystore and password
-                certAndKey = addKeyAndCertToKeyStore(subject.commonName(),
-                        Base64.getDecoder().decode(secretEntryDataForPod(secret, podName, SecretEntry.KEY)),
-                        Base64.getDecoder().decode(secretEntryDataForPod(secret, podName, SecretEntry.CRT)));
-            }
+        for (NodeRef node : nodes)  {
+            String podName = node.podName();
+            Subject subject = subjectFn.apply(node);
+
+            if (!this.certRenewed() // No CA renewal is happening
+                    && secret != null && secret.getData() != null // Secret exists and has some data
+                    && secretEntryExists(secret, podName, SecretEntry.CRT) // The secret has the public key for this pod
+                    && secretEntryExists(secret, podName, SecretEntry.KEY) // The secret has the private key for this pod
+            )   {
+                // A certificate for this node already exists, so we will try to reuse it
+                LOGGER.debugCr(reconciliation, "Certificate for node {} already exists", node);
+
+                CertAndKey certAndKey;
+
+                if (isNewVersion(secret, podName)) {
+                    certAndKey = asCertAndKey(secret, podName);
+                } else {
+                    // coming from an older operator version, the secret exists but without keystore and password
+                    certAndKey = addKeyAndCertToKeyStore(subject.commonName(),
+                            Base64.getDecoder().decode(secretEntryDataForPod(secret, podName, SecretEntry.KEY)),
+                            Base64.getDecoder().decode(secretEntryDataForPod(secret, podName, SecretEntry.CRT)));
```

Comment:
I was wondering if at some point we could get rid of this code but we should check the older version which still needs it.